lib/Conversion/ImportVerilog/CMakeLists.txt

```diff
@@ -0,0 +1,38 @@
+# slang uses exceptions
+set(LLVM_REQUIRES_EH ON)
+set(LLVM_REQUIRES_RTTI ON)
```

Comment:
> We manage to make this work in the Python bindings using pybind11, which uses exceptions and RTTI. I haven't followed exactly how this is done, but there are some helpers in this file that mess with the settings: https://github.com/llvm/llvm-project/blob/b52fe2df9178022f146fe1776b9a9fd7f4b29bb4/mlir/cmake/modules/AddMLIRPython.cmake#L639-L647

Unfortunately, all the pybind11 code goes through the CAPI. Max's stuff is only guaranteed to work with LLVM compiled with EH and RTTI. Unfortunately, C++ compilers (apparently) don't include those information in the code objects output, so stuff can go terribly wrong at runtime. (e.g. a shared class layout changes based on RTTI, so the non-RTTI object code corrupts the class instance produced by the RTTI object code.)

Reply:
Ah yes, we go through the C API. I guess going through the C API is the sure bet. But I do know triton uses MLIR C++ APIs directly alongside pybind11: https://github.com/openai/triton/blob/main/python/src/ir.cc. I'm not sure how it's being compiled, but I was under the impression pybind11 uses exceptions and RTTI so maybe it is possible? I don't see anything about RTTI and exceptions here, so maybe this is not worth considering: https://github.com/openai/triton/blob/main/python/setup.py

Reply:
I remember looking into this when I was told about Triton. They may or may not be compiling LLVM with EH/RTTI. Linkers aren't required to check for compatibility, so it may compile and work fine 99.999% of the time on some platform/compiler configurations, but... And I have no idea what dynamic linking does, but I'm working under the assumption (which is consistent with Stella's experience/scars) that dynamically linking C++ is generally not safe if the two shared objects have been compiled with radically different flags (e.g. EH/RTTI). Either way, I'm now of the opinion that just because something "works" doesn't mean it'll actually work. Obscure linking issues causing memory corruption are _very_ difficult to debug, and I don't want to land there.

So I hope Triton is compiling LLVM with RTTI/EH.

Reply:
> This only sets this for these libraries, right?

Yeah these should be limited to the current CMakeLists file.

> The simplest option is to require building LLVM/MLIR with these enabled

That would definitely be an option. I'm not quite sure what to do here. When I did some digging regarding this, it looked like carefully catching exceptions would allow us to bridge between EH/non-EH code. But the stuff @teqdruid points out is concerning, and suggests otherwise.

Do we want to pay the cost (no idea how much) of EH/RTTI? We could also just roll with the current version until we see an actual breakage (pretty sure that'll be msvc :wink:), and at that point fall back to enabling EH/RTTI for the entire project.

I don't think going through a C API is realistic here. We can definitely make the call into ImportVerilog through a C function, but we'd still be building up a C++ IR on the EH/RTTI side, whose memory layout would still differ from what the non-EH/non-RTTI side expects.

Reply:
The MLIR CAPI is all about building up IR. Not having the generated facade classes is the "only" thing you don't get -- which is for sure a real pain.

> We could also just roll with the current version until we see an actual breakage

Dangerous IMO.

> fall back to enabling EH/RTTI for the entire project

This should only be required for `ImportVerilog` builds.

Reply:
rtti/no-rtti should be safe is exceptions don't cross the lines (though I think that is handled these days) and if class hierarchies don't intersect and if casts are done with the right implementation.  In practice I agree this is fragile and likely painful.

Instead of changing llvm mode or using the clang API, can we wrap slang?  A slang wrapper with rtti used by llvm (no-rtti) can be made safe with a few rules.

Can we clean slang?  A cursory glance seems to say the only dynamic_cast is in boost_unordered.  Switching to std::unordered_map might be enough.  It looks like slang already has a non-rtti fallback for type_index.  Looking more, the entire rtti dependency seems to be for slang::Bag.  (std::tuple also has by-type lookup, without rtti).  Clearly effort has been put in to get this to compile with llvm.

I would very much prefer not to make circt have a dependency on non-standard llvm builds.

Reply:
I would very much prefer not to make circt have a dependency on non-standard llvm builds as this will make mainlining circt nearly impossible.

Reply:
Yeah. I'm very reluctant to funnel things through the C API just because some C++ compilers have troublesome ABI compatibility across build options. Especially since all the frontend does is build up IR, which is literally what all the builders, lowering tools, and other stuff in MLIR is about.

Reply:
> Especially since all the frontend does is build up IR, which is literally what all the builders, lowering tools, and other stuff in MLIR is about.

Same is true of everything else which uses the CAPI! I agree it sucks... but I've been dealing with that same pain via Python / PyCDE for a loooong time. (Though the ODS Python gen does a lot of the heavy lifting.)

The far better solution is cleaning slang! I'm still not sure that EH and non-EH code can be mixed safely -- even if the exceptions are caught inside the EH code -- but I think that's far more possible than RTTI.

Reply:
> @darthscsi Instead of changing llvm mode or using the clang API, can we wrap slang? A slang wrapper with rtti used by llvm (no-rtti) can be made safe with a few rules.

That's what `ImportVerilog` currently is. It's built with RTTI and EH, but wraps all calls into slang in a `try { ... }` that catches exceptions and prevents them from escaping into non-EH LLVM code.

Reply:
Another option would be to try to use Slang 4 or 5, which can be built in a non-RTTI/non-EH way. But those versions have stricter compiler requirements: C++20 and GCC 11, clang 16, Xcode 15. Last time I tried this didn't work with our CI images. But that could be a way forward.

Reply:
> Another option would be to try to use Slang 4 or 5, which can be built in a non-RTTI/non-EH way. But those versions have stricter compiler requirements: C++20 and GCC 11, clang 16, Xcode 15. Last time I tried this didn't work with our CI images. But that could be a way forward.

Since Slang 4/5 has gone through the effort, this is the option I prefer. I can help update the Linux images. Should be simple given that they're all docker images.

Reply:
That sounds great! Thanks for taking care of the Linux docker images -- I'm not familiar with how exactly we've set those up.